src/main/java/com/google/devtools/build/lib/util/Fingerprint.java

```diff
@@ -84,6 +79,14 @@ public Fingerprint() {
     this(DigestHashFunction.SHA256);
   }
 
+  private void reset() {
+    hasher = hashFunction.newHasher();
+    // This is a lot of indirection, but CodedOutputStream does a reasonable job of converting
+    // strings to bytes without creating a whole bunch of garbage, which pays off.
+    codedOut =
```

Comment:
this shouldn't be happening on every reset--the codedOut.flush() calls should be good enough to ensure that the stream is empty before the next hash starts, if i'm not misreading.

Reply:
I think the issue is that codedOut, because it was created with a hasher, needs to be recreated with a new hasher (since hashers cannot be re-used). Maybe there's a more efficient way to do this or I'm misunderstanding?

Reply:
ahh crap, i failed to read that.  alright, i think the answer should be the same as the SHA-with-`MessageDigest` vs SHA-with-`HashFunction` stuff above, yeah? not really any other analysis necessary.